Add Encryption and Compression for Local File Spills

pettyjamesm opened the pull request:
This PR supersedes https://github.com/prestodb/presto/pull/8801, which originally implemented spill encryption using bouncy-castle. This PR implements the the encrypt / decrypt operations using the platform-provided JCE implementation which has much better performance on most JDK implementations thanks to AES-NI intrinsics.

The behavior is controlled via `FeaturesConfig("experimental.spill-encryption-enabled")` (default: off). When enabled, spill data is encrypted with a randomly generated key per `FileSingleStreamSpiller` and destroyed by `FileSingleStreamSpiller#close`.

mbasmanova commented:
@pettyjamesm James, GitHub says `This branch has conflicts that must be resolved`. Could you rebase and resolve conflicts?

pettyjamesm commented:
@mbasmanova will do, looks like the conflicts were introduced within a few hours of me opening the PR. I’ll rebase and address them tomorrow.

pettyjamesm commented:
@mbasmanova I rebased and resolved conflicts, but the tests failed this time around in presto-verifier. Looking through the log, it looks like presto verifier failed to start mysql correctly, which shouldn't have anything to do with my changes. From the logs:

```
2019-04-09T12:16:47.699-0500	INFO	testing-mysql-server-0	stdout	2019-04-09T17:16:47.698903Z 0 [ERROR] [MY-010262] [Server] Can't start server: Bind on TCP/IP port: Address already in use
2019-04-09T17:16:47.698922Z 0 [ERROR] [MY-010257] [Server] Do you already have another mysqld server running on port: 34920 ?
2019-04-09T17:16:47.698938Z 0 [ERROR] [MY-010119] [Server] Aborting
```

Is it possible someone could trigger a retry of the build job? Looks like it was probably just a race condition.

mbasmanova commented:
@pettyjamesm Restarted the build.

pettyjamesm commented:
@mbasmanova thanks for restarting the build, it passed on the second attempt.

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/PageCodecMarker.java:
```diff
@@ -0,0 +1,77 @@
+        return (byte) marker.mask;
+    }
+
+    public static byte create(PageCodecMarker marker, PageCodecMarker... markers)
```
this method seems be to unused

pettyjamesm replied:
Will remove.

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/PageCodecMarker.java:
```diff
@@ -0,0 +1,77 @@
+        return 0;
+    }
+
+    public static byte create(PageCodecMarker marker)
```
this method is a bit strange; it takes an object and returns a primitive type; the name create would suggest the opposite: take some primitives and construct an object; also, this method doesn't seem to be used in production code. Would you clarify its purpose?

pettyjamesm replied:
Sure, I'll add an comment block. Maybe the method would be better named `of` instead of `create`?

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/PageCodecMarker.java:
```diff
@@ -0,0 +1,77 @@
+
+    private final int mask;
+
+    PageCodecMarker(int flagBitPosition)
```
I'd rename flagBitPosition to bit for clarity

pettyjamesm replied:
Will change.

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/PageCodecMarker.java:
```diff
@@ -0,0 +1,77 @@
+    PageCodecMarker(int flagBitPosition)
+    {
+        if (flagBitPosition <= 0 || flagBitPosition > 8) {
+            throw new IllegalArgumentException("PageCodecMarker flagBitPosition must be between 1 and 8. Found: " + flagBitPosition);
```
use checkArgument: `checkArgument(bit >= 1 && bit <= 8, "bit must be between 1 and 8; found " + bit);`

pettyjamesm replied:
Will change to use `checkArgument `

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/PagesSerde.java:
```diff
@@ -56,31 +54,33 @@ public SerializedPage serialize(Page page)
 
         if (!compressor.isPresent()) {
-            return new SerializedPage(serializationBuffer.slice(), UNCOMPRESSED, page.getPositionCount(), serializationBuffer.size());
+            return new SerializedPage(serializationBuffer.slice(), markings, page.getPositionCount(), serializationBuffer.size());
```
PageCodecMarker.none() for clarity

pettyjamesm replied:
Will change.

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/PagesSerde.java:
```diff
@@ -56,31 +54,33 @@ public SerializedPage serialize(Page page)
 
         if (((1.0 * actualCompressedLength) / serializationBuffer.size()) > MINIMUM_COMPRESSION_RATIO) {
-            return new SerializedPage(serializationBuffer.slice(), UNCOMPRESSED, page.getPositionCount(), serializationBuffer.size());
+            return new SerializedPage(serializationBuffer.slice(), markings, page.getPositionCount(), serializationBuffer.size());
```
PageCodecMarker.none() for clarity

pettyjamesm replied:
Will Change.

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/PagesSerde.java:
```diff
@@ -56,31 +54,33 @@ public SerializedPage serialize(Page page)
-            return new SerializedPage(serializationBuffer.slice(), UNCOMPRESSED, page.getPositionCount(), serializationBuffer.size());
+            return new SerializedPage(serializationBuffer.slice(), markings, page.getPositionCount(), serializationBuffer.size());
+        }
+        else {
```
else after return is not needed

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/PagesSerde.java:
```diff
@@ -56,31 +54,33 @@ public SerializedPage serialize(Page page)
+        else {
+            return new SerializedPage(
+                    Slices.copyOf(Slices.wrappedBuffer(compressionBuffer, 0, actualCompressedLength)),
+                    PageCodecMarker.COMPRESSED.mark(markings),
```
`COMPRESSED.get()` would be clearer

pettyjamesm replied:
Rather, `PageCodecMarker.of(COMPRESSED)`. Will change.

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/PageCodecMarker.java:
```diff
@@ -0,0 +1,77 @@
+        return markings;
+    }
+
+    public static String toSummaryString(byte marks)
```
I'm wondering if it would be easier to read the output if it contains markers that are set and doesn't include all of them:

```
    public static String toSummaryString(byte markers)
    {
        if (markers == none()) {
            return "NONE";
        }
        
        return Arrays.stream(values())
                .filter(codec -> codec.isSet(markers))
                .map(codec -> codec.name())
                .collect(Collectors.joining(", "));
    }

NONE
COMPRESSED
ENCRYPTED
COMPRESSED, ENCRYPTED
```

pettyjamesm replied:
Makes sense, will change to that.

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/SpillCipher.java:
```diff
@@ -0,0 +1,52 @@
+     * Encrypts the contents of the input ByteBuffer into a new ByteBuffer instance. The returned instance will
+     * be at position 0 and ready to read.
+     */
+    ByteBuffer encrypt(ByteBuffer plainText);
```
The contents of the buffer may not be text, hence, consider renaming plainText to data.

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/SpillCipher.java:
```diff
@@ -0,0 +1,52 @@
+     * Decrypts the given {@link ByteBuffer} contents and returns them in a new {@link ByteBuffer}. The returned
+     * instance will be at position 0 and ready to read.
+     */
+    ByteBuffer decrypt(ByteBuffer plainText);
```
plainText -> encryptedData

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/AESSpillCipher.java:
```diff
@@ -0,0 +1,219 @@
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+final class AESSpillCipher
```
AESSpillCipher -> AesSpillCipher

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/AESSpillCipher.java:
```diff
@@ -0,0 +1,219 @@
+    }
+
+    @Override
+    public ByteBuffer decrypt(ByteBuffer cipherText)
```
cipherText -> encryptedData

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/AESSpillCipher.java:
```diff
@@ -0,0 +1,219 @@
+    {
+        int length = cipherText.remaining();
+        if (length < ivBytes) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, format("cipherText too small to contain IV. Minimum: %s, Actual: %s", ivBytes, length));
```
Perhaps, clarify the error message and spell out IV: `Encrypted data size is too small: %s. It must be at least the size of the initialization vector: %s.`

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/AESSpillCipher.java:
```diff
@@ -0,0 +1,219 @@
+            return output;
+        }
+        catch (GeneralSecurityException e) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, "Failure during decrypt", e);
```
perhaps, `"Cannot decrypt previously encrypted data: " + e.getMessage()`

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/AESSpillCipher.java:
```diff
@@ -0,0 +1,219 @@
+    }
+
+    @Override
+    public byte[] decrypt(byte[] cipherText)
```
cipherText -> encryptedData

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/AESSpillCipher.java:
```diff
@@ -0,0 +1,219 @@
+    }
+
+    @Override
+    public ByteBuffer encrypt(ByteBuffer plainText)
```
plainText -> data

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/AESSpillCipher.java:
```diff
@@ -0,0 +1,219 @@
+    }
+
+    @Override
+    public byte[] encrypt(byte[] plainText)
```
plainText -> data

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/AESSpillCipher.java:
```diff
@@ -0,0 +1,219 @@
+            return output;
+        }
+        catch (GeneralSecurityException e) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, "Failure during encrypt", e);
```
perhaps, "Failed to encrypt data: " + e.getMessage()

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/AESSpillCipher.java:
```diff
@@ -0,0 +1,219 @@
+        if (result.hasArray() && result.arrayOffset() == 0 && result.capacity() == result.remaining()) {
+            return result.array();
+        }
+        else {
```
else after return is unnecessary

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/AESSpillCipher.java:
```diff
@@ -0,0 +1,219 @@
+     * their buffer. This is expected during wrapped encrypt / decrypt buffers and those buffers are
+     * not externally visible (and therefore safe to unwrap without copies)
+     */
+    private static byte[] unwrapOrCopyToByteArray(ByteBuffer result)
```
result -> buffer

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/AESSpillCipher.java:
```diff
@@ -0,0 +1,219 @@
+    @Override
+    public boolean isDestroyed()
+    {
+        SecretKeySpec key = this.key;
```
- this variable is not needed: `return this.key == null || this.key.isDestroyed();`
- how can the key become destroyed? I don't see any calls to key.destroy().

pettyjamesm replied:
To address the first point, the `SecretKey` instance stored in a local variable before doing the checks against it because `AesSpillCipher.destroy()` sets `this.key = null;`. In the case of `AesSpillCipher.isDestroyed()` storing it in a temporary variable is just being paranoid- however in your later comment about `verifyKeyNotDestroyed()` the value of creating the local variable and returning it is to avoid a race condition between destruction and key usage that would raise an unhelpful `NullPointerException` if triggered instead of a `PrestoException` explaining the cipher was already destroyed.

On point 2... this is kind of a weird edge case in the JCE libraries. The `SecretKey` interface extends [Destroyable](https://hg.openjdk.java.net/jdk8u/jdk8u/jdk/file/cecd70d27b27/src/share/classes/javax/security/auth/Destroyable.java), but neither the `SecretKey` interface nor `SecretKeySpec` class overrides the `destroy()` method as of JDK 8, so the interface default implementation is used which unconditionally throws `DestroyFailedException`.

That left me with the option of either calling `key.destroy()` inside of `AesSpillCipher.destroy()` but catching and ignoring the `DestroyFailedException` (maybe someday the method will be actually implemented, right?) or not calling it at all. I chose to just not call the method, but left the `.isDestroyed()` check to be slightly defensive against future refactors. I'm fine with removing the check to `key.isDestroyed()` or even reintroducing the call to `key.destroy()` and ignoring the exception. Let me know what you think is best.

mbasmanova replied:
@pettyjamesm James, thanks for the detailed explanation. I'm wondering now why do you need `destroy` method at all? Why not just let GC claim the object eventually? If there is a good reason to close the instance and prevent its usage after that, then it might be simpler to add `private boolean closed;` member variable and add `checkState(!closed, "...")` everywhere the key is used.

pettyjamesm replied:
For our use-cases, "destroying" the key after it's no longer necessary to use is desirable. It's an insurance policy against keys living longer than intended or being accidentally reused outside of the intended "one-per-spill-file" lifespan. By assigning the key to `null` in `AesSpillCipher#destroy()` we're both accomplishing the same functionality as `checkState(!closed, ...)` and making sure the secret key is eligible for reclamation by the GC even if the `AesSpillCipher` remains reachable for some reason.

I think leaving the check in for `key.isDestroyed()`, while paranoid- is probably preferable. I can see a world where users that enable spill encryption may want to manage key lifecycles differently than one-per-spill-file which might introduce the possibility of key originating and being destroyed outside of the `AesSpillCipher` at which point failing with a descriptive message is preferable to whatever exception would bubble from attempting to use a destroyed key. If you agree, I'm inclined to leave in the `key == null || key.isDestroyed()` check.

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/AESSpillCipher.java:
```diff
@@ -0,0 +1,219 @@
+
+    private SecretKeySpec verifyKeyNotDestroyed()
+    {
+        SecretKeySpec key = this.key;
```
This variable is not needed as well as the `else` below:

```
        if (key == null || key.isDestroyed()) {
            throw new PrestoException(GENERIC_INTERNAL_ERROR, "SpillCipher already destroyed");
        }
        return key;
```

Also, this method could just return void and the callers could simply use this.key:

```
    private void verifyKeyNotDestroyed()
    {
        if (key == null || key.isDestroyed()) {
            throw new PrestoException(GENERIC_INTERNAL_ERROR, "SpillCipher already destroyed");
        }
    }

    private Cipher createEncryptCipher()
    {
        verifyKeyNotDestroyed();
        Cipher cipher = createUninitializedCipher();
        try {
            cipher.init(Cipher.ENCRYPT_MODE, key);
            return cipher;
        }
        catch (GeneralSecurityException e) {
            throw new PrestoException(GENERIC_INTERNAL_ERROR, "Failed to initialize cipher for encryption", e);
        }
    }
```

pettyjamesm replied:
See [my explanation above](https://github.com/prestodb/presto/pull/12617#discussion_r277841598) for the rationale behind using a local variable and returning it from `verifyKeyNotDestroyed()`.

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/AESSpillCipher.java:
```diff
@@ -0,0 +1,219 @@
+    {
+        SecretKeySpec key = this.key;
+        if (key == null || key.isDestroyed()) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, "SpillCipher already destroyed");
```
"Spill cipher already destroyed"

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/AESSpillCipher.java:
```diff
@@ -0,0 +1,219 @@
+            return cipher;
+        }
+        catch (GeneralSecurityException e) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, "Failed to initialize cipher for encryption", e);
```
"Failed to initialize spill cipher for encryption: " + e.getMessage()

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/AESSpillCipher.java:
```diff
@@ -0,0 +1,219 @@
+            return cipher;
+        }
+        catch (GeneralSecurityException e) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, "Failed to initialize cipher for decryption", e);
```
"Failed to initialize spill cipher for decryption: " + e.getMessage()

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/AESSpillCipher.java:
```diff
@@ -0,0 +1,219 @@
+            return Cipher.getInstance("AES/CBC/PKCS5Padding");
+        }
+        catch (GeneralSecurityException e) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, "Failed to create cipher", e);
```
"Failed to create spill cipher: " + e.getMessage()

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/AESSpillCipher.java:
```diff
@@ -0,0 +1,219 @@
+    private static Cipher createUninitializedCipher()
+    {
+        try {
+            return Cipher.getInstance("AES/CBC/PKCS5Padding");
```
Would you add a constant for this value?

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/AESSpillCipher.java:
```diff
@@ -0,0 +1,219 @@
+        }
+    }
+
+    private static SecretKeySpec generateNewSecretKey(int keyBits)
```
keyBits -> numBits ?

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/AESSpillCipher.java:
```diff
@@ -0,0 +1,219 @@
+            if (key instanceof SecretKeySpec) {
+                return (SecretKeySpec) key;
+            }
+            else {
```
this `else` is not needed

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/AESSpillCipher.java:
```diff
@@ -0,0 +1,219 @@
+                return new SecretKeySpec(key.getEncoded(), key.getAlgorithm());
+            }
+        }
+        catch (NoSuchAlgorithmException nsa) {
```
nsa -> e

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/PagesSerde.java:
```diff
@@ -39,56 +41,81 @@
     private final BlockEncodingSerde blockEncodingSerde;
-    private final Optional<Compressor> compressor;
-    private final Optional<Decompressor> decompressor;
+    @Nullable private final Compressor compressor;
```
Why is this change?

pettyjamesm replied:
To avoid unnecessary allocations and indirection. Since spilling can be invoked as a result of memory revocation and one `PagesSerde` is created per `SingleStreamSpiller`, I thought it better to use nullable references instead, but I can revert it back if you prefer. I have no data to support this change being measurable in reducing memory pressure or improving performance.

mbasmanova replied:
@pettyjamesm I think these kind of optimizations make sense only when operation is happening for each row (e.g. number of operations is proportional to the amount of data being processed). It seems like this is not the case here. It is always hard to tell without a benchmark. If you feel like this is a necessary optimization, add a separate commit which applies it to compressor and decompressor before encryption commits.

pettyjamesm replied:
I'll back the change out and use `Optional`

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/PagesSerde.java:
```diff
@@ -39,56 +41,81 @@
+            ByteBuffer compressionBuffer = ByteBuffer.allocate(compressor.maxCompressedLength(uncompressedSize));
+            compressor.compress(slice.toByteBuffer(), compressionBuffer);
+            compressionBuffer.flip();
+            if ((((double) compressionBuffer.remaining()) / ((double) uncompressedSize)) <= MINIMUM_COMPRESSION_RATIO) {
```
second cast to double is not necessary: `if ((((double) compressionBuffer.remaining()) / uncompressedSize) <= MINIMUM_COMPRESSION_RATIO) {`

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/PagesSerde.java:
```diff
@@ -39,56 +41,81 @@
+        Slice slice = serializedPage.getSlice();
+
+        if (PageCodecMarker.ENCRYPTED.check(serializedPage.getPageCodecMarkings())) {
+            checkState(spillCipher != null, "Page is marked encrypted, but no spillCipher is present");
```
"Page is encrypted, but spill cipher is missing"

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/PagesSerde.java:
```diff
@@ -39,56 +41,81 @@
-        int actualUncompressedSize = decompressor.get().decompress(serializedPage.getSlice().getBytes(), 0, serializedPage.getSlice().length(), decompressed, 0, uncompressedSize);
-        checkState(uncompressedSize == actualUncompressedSize);
+        if (PageCodecMarker.COMPRESSED.check(serializedPage.getPageCodecMarkings())) {
+            checkState(decompressor != null, "decompressor is not present, but page is marked as compressed");
```
"Page is compressed, but decompressor is missing"

pettyjamesm commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/PageCodecMarker.java:
```diff
@@ -0,0 +1,77 @@
+        this.mask = (1 << (flagBitPosition - 1));
+    }
+
+    public boolean check(byte value)
```
Will change to `set` / `isSet`.

pettyjamesm commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/PagesSerde.java:
```diff
@@ -56,31 +54,33 @@ public SerializedPage serialize(Page page)
     {
         SliceOutput serializationBuffer = new DynamicSliceOutput(toIntExact((page.getSizeInBytes() + Integer.BYTES))); // block length is an int
         writeRawPage(page, serializationBuffer, blockEncodingSerde);
+        byte markings = PageCodecMarker.none();
```
Will change.

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/PagesSerdeFactory.java:
```diff
@@ -25,19 +27,42 @@
+    }
+
+    /**
+     * Creates a {@link PagesSerde} instance for use in spilling to disk, which will be encrypted if {@link PagesSerdeFactory#spillEncryptionEnabled}
```
Creates a {@link PagesSerde} instance for use in spilling to disk. The data will be encrypted if {@link PagesSerdeFactory#spillEncryptionEnabled} is set.

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/AESSpillCipher.java:
```diff
@@ -0,0 +1,219 @@
+{
+    public enum Mode
+    {
+        AES_128(128),
```
128-bit keys seem to not be used in production. What's the reason for including them?

pettyjamesm replied:
This is there for a couple reasons, none of them great:
- AES 128 can be (just barely) faster than AES 256 and theoretically someone might want to opt into that behavior (although I did not include a config switch for that in this PR)
- The same `AesSpillCipher` implementation works for both, but 256-bit AES isn't universally available. Oracle JDK releases before 8u161 don't support 256-bit AES out of the box and required you to manually install the "Unlimited Strength Jurisdiction Policy" for each version installed, whereas 128-bit AES was supported out of the box. I figured at least some new developers checking out presto and running tests locally might bump into that issue.

Let me know if you'd like to see it behind a different config flag, removed, or fine with it as-is.

mbasmanova replied:
@pettyjamesm James, thanks for explaining. For simplicity sake, let's remove this option. I suggest we explain any gotchas you know of in the docs. BTW, let's update Spill to Disk documentation to describe this new functionality: https://prestodb.github.io/docs/current/admin/spill.html

pettyjamesm replied:
This is removed in the latest revision.

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/PagesSerdeFactory.java:
```diff
@@ -25,19 +27,42 @@
+     * Creates a {@link PagesSerde} instance for use in spilling to disk, which will be encrypted if {@link PagesSerdeFactory#spillEncryptionEnabled}
+     * is set.
+     */
+    public PagesSerde createSpillerPagesSerde()
```
perhaps, rename to createPagesSerdeForSpill

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/PagesSerdeFactory.java:
```diff
@@ -25,19 +27,42 @@
+     */
+    public PagesSerde createSpillerPagesSerde()
+    {
+        return createPagesSerdeInternal(spillEncryptionEnabled ? Optional.of(new AESSpillCipher(AESSpillCipher.Mode.AES_256)) : Optional.empty());
```
consider introducing temp variable for readability and consistency with createPagesSerdeInternal:

```
        Optional<SpillCipher> spillCipher = Optional.empty();
        if (spillEncryptionEnabled) {
            spillCipher = Optional.of(new AesSpillCipher(AesSpillCipher.Mode.AES_256));
        }
        return createPagesSerdeInternal(spillCipher);
```

pettyjamesm commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/SerializedPage.java:
```diff
@@ -16,31 +16,32 @@
-        checkArgument(compression == UNCOMPRESSED || uncompressedSizeInBytes > slice.length(), "compressed size must be smaller than uncompressed size when compressed");
-        checkArgument(compression == COMPRESSED || uncompressedSizeInBytes == slice.length(), "uncompressed size must be equal to slice length when uncompressed");
         this.uncompressedSizeInBytes = uncompressedSizeInBytes;
+        if (PageCodecMarker.COMPRESSED.check(pageCodecMarkings)) {
```
Will change to use static imports.

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java:
```diff
@@ -587,6 +588,18 @@ public FeaturesConfig setSpillEnabled(boolean spillEnabled)
+        return spillEncryptionEnabled;
+    }
+
+    @Config("experimental.spill-encryption-enabled")
```
Why experimental? What's the plan to make it mainstream?

pettyjamesm replied:
I was trying to match the convention for other spill-to-disk related properties and the flag to enable / disable spilling altogether is also marked `experimental`.

pettyjamesm commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/SerializedPage.java:
```diff
@@ -16,31 +16,32 @@
+    private final byte pageCodecMarkings;
 
-    public SerializedPage(Slice slice, PageCompression compression, int positionCount, int uncompressedSizeInBytes)
+    public SerializedPage(Slice slice, byte pageCodecMarkings, int positionCount, int uncompressedSizeInBytes)
```
I'm not sure, since we aren't really identifying a single "Codec" but instead just boolean properties of the way the page was encoded / should be decoded. Would `PageMarker` be better? I agree that `PageCodecMarker` is a little too verbose, but I failed to come up with something better myself.

pettyjamesm commented:
@mbasmanova Thanks for the review! I've rebased on top of latest master and resolved comments that I addressed based on your feedback. I've replied to the remaining unresolved comments with context and possible implementation choices.

abajenov commented:
A did a quick pass, and have a few initial thoughts:
- I would remove the ability to use 128 bit cipher. If the platform doesn't support 256 bit, then don't allow enabling encryption imo.
- It might be worth adding / validating a version alongside the ciphertext you're storing.  This would help support rotation to new ciphers / encryption schemes later.
- Doing compression before encryption is in some cases not safe. If there is an oracle allowing a user to modify part of the original plaintext, the mere size of the final ciphertext can leak data about the rest of the plaintext. Not sure if this applies here, but something to think about.

pettyjamesm commented:
Thanks for the feedback @abajenov :

>     * I would remove the ability to use 128 bit cipher. If the platform doesn't support 256 bit, then don't allow enabling encryption imo.

I agree, I'll remove AES 128 support.

>     * It might be worth adding / validating a version alongside the ciphertext you're storing.  This would help support rotation to new ciphers / encryption schemes later.

I don't think we need to worry about that for now, since these files are temporary and only exist to offload data from memory for short parts of query execution. No historical data is meant to be read except using the key and cipher that created the file, and even then only for the duration of the query creating those temporary files. The key itself is generated per temporary file and discarded when the file is no longer needed.

>     * Doing compression before encryption is in some cases not safe. If there is an oracle allowing a user to modify part of the original plaintext, the mere size of the final ciphertext can leak data about the rest of the plaintext. Not sure if this applies here, but something to think about.

Good call out. I don't think that's a big deal for this use-case since this is meant to protect query data on disk from being recovered. If I understand correctly, using compressed size as an oracle would require the ability to manipulate the query data itself. To me, that seems to imply that the attacker already has access to the query data but also would require repeated attempts and local machine access to inspect file size. Is that correct?

Either way, compression and encryption currently don't interact since spill to disk has hard-coded compression to be disabled. That said, if we want to prevent encryption / compression from coexisting then we can refactor the code to isolate the two from each other and reduce the likelihood someone adds the capability later on without realizing it. After all, I definitely wrote the code here with the intention of being able to support that.

abajenov commented:
Makes sense. If this data is truly ephemeral, I'm ok that you don't add versioning. From experience though it does help with debugging. Adding some structure to the ciphertext to make it clear how it was generated is sometimes helpful, otherwise you're just looking at a blob of random-looking data. 

You are right about compression. It is only a problem if the user has control of part of the data that is being encrypted. Controlling part of the encrypted data may let him infer things about the rest of the encrypted data. If we can guarantee that this kind of thing won't happen, then we can compress. Unfortunately I don't know too much about what data is being spilled here, so I'm just bringing it up in case you think it may be a problem. :)

pettyjamesm commented:
@abajenov Got it, thanks for the clarification. The data we're talking about here is essentially intermediate data being processed during query execution, and the point of adding encryption is to avoid writing plaintext intermediate data to persistent storage that could potentially be recovered from that storage later.

I want to make sure I understand the specifics of the attack you're describing so I can add details to the documentation. Is it correct to say:

If an attacker:
- has control over some part of the data being compressed and encrypted (eg: some data processed by the a query execution)
- can trigger processing with different variations of the attacker controlled data
- can determine relative size of the compressed + encrypted data between variations (in this case, that implies access to the temporary file)

the attacker might be able to infer aspects of other data co-processed with the attacker controlled portion?

pettyjamesm commented:
@mbasmanova I've addressed all the comments you left previously and done some refactors to address the `SpillCipher` lifecycle awkwardness. `PagesSerde` no longer needs a `destroy()` method since the file spiller can manage the `SpillCipher` directly.

I also moved the configuration flag to `NodeSpillConfig` and after doing so, it was trivial to also add a flag for spill to disk compression. I'll want to make sure I add a good summary description of the risks involved with combining compression and encryption, so I'll wait for @abajenov to confirm my description above is correct and add that to the documentation.

Lots of the changes originally present to pass `spillEncryptionEnabled` into `PagesSerdeFactory` are no longer necessary and evaporated from the PR noise.

abajenov commented:
@pettyjamesm yup that's exactly it. Nice summary. :)

mbasmanova commented:
@pettyjamesm James, thanks for addressing comments. I'm going to take another pass. In the meantime, would you update commit message to wrap them at 78 characters?

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/PagesSerde.java:
```diff
@@ -41,50 +45,73 @@
-        if (!compressor.isPresent()) {
-            return new SerializedPage(serializationBuffer.slice(), PageCodecMarker.none(), page.getPositionCount(), serializationBuffer.size());
+        if (compressor.isPresent()) {
+            Compressor c = compressor.get();
```
c -> compressor; also, consider inlining this variable as it is used only in two places

pettyjamesm replied:
Sadly, compressor is already the name of the `Optional<Compressor>`. I'll just repeat the call to `.get()`.

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/PagesSerdeFactory.java:
```diff
@@ -34,10 +37,27 @@ public PagesSerdeFactory(BlockEncodingSerde blockEncodingSerde, boolean compress
         }
 
-        return new PagesSerde(blockEncodingSerde, Optional.empty(), Optional.empty());
+        return new PagesSerde(blockEncodingSerde, compressor, decompressor, spillCipher);
```
The original code style in this method is better:

```
    private PagesSerde createPagesSerdeInternal(Optional<SpillCipher> spillCipher)
    {
        if (compressionEnabled) {
            return new PagesSerde(blockEncodingSerde, Optional.of(new Lz4Compressor()), Optional.of(new Lz4Decompressor()), spillCipher);
        }

        return new PagesSerde(blockEncodingSerde, Optional.empty(), Optional.empty(), spillCipher);
    }
```

pettyjamesm replied:
Sure, will change to this implementation.

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/spiller/AesSpillCipher.java:
```diff
@@ -0,0 +1,190 @@
+        if (key == null || key.isDestroyed()) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, "Spill cipher already destroyed");
+        }
+        else {
```
drop unnecessary `else`.

```
    private SecretKey verifyKeyNotDestroyed()
    {
        SecretKey key = this.key;
        if (key == null || key.isDestroyed()) {
            throw new PrestoException(GENERIC_INTERNAL_ERROR, "Spill cipher already destroyed");
        }
        return key;
    }
```

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/spiller/AesSpillCipher.java:
```diff
@@ -0,0 +1,190 @@
+    @Override
+    public void destroy()
+    {
+        this.key = null;
```
Add a comment explaining why this is useful. On the surface, it doesn't seem necessary.

pettyjamesm replied:
Will add a comment to the method.

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/spiller/FileSingleStreamSpillerFactory.java:
```diff
@@ -83,9 +88,11 @@ public FileSingleStreamSpillerFactory(
+            boolean spillEncryptionEnabled)
     {
-        this.serdeFactory = new PagesSerdeFactory(requireNonNull(blockEncodingSerde, "blockEncodingSerde is null"), false);
+        this.serdeFactory = new PagesSerdeFactory(requireNonNull(blockEncodingSerde, "blockEncodingSerde is null"), spillCompressionEnabled);
```
Would you extract logic to allow enabling/disabling compression for spill into a separate commit?

pettyjamesm replied:
Will do.

mbasmanova commented on presto-main/src/test/java/com/facebook/presto/spiller/TestSpillCipher.java:
```diff
@@ -0,0 +1,80 @@
+public class TestSpillCipher
+{
+    @Test
+    public void testAESSpillCipher()
```
- I'd inline all the methods that are used only once for clarity and simplicity.
- I'd add logic to verify error messages

```
    @Test
    public void test()
    {
        AesSpillCipher spillCipher = new AesSpillCipher();
        // test [0, 257] buffer sizes to check off all padding cases
        for (int size = 0; size <= 257; size++) {
            byte[] data = randomBytes(size);
            assertEquals(data, spillCipher.decrypt(spillCipher.encrypt(data)));
            assertEquals(ByteBuffer.wrap(data), spillCipher.decrypt(spillCipher.encrypt(ByteBuffer.wrap(data))));
        }

        // verify that initialization vector is not re-used
        assertNotEquals(spillCipher.encrypt(new byte[0]), spillCipher.encrypt(new byte[0]), "IV values must not be reused");

        // cleanup
        assertFalse(spillCipher.isDestroyed());
        byte[] encrypted = spillCipher.encrypt(randomBytes(1));
        spillCipher.destroy();
        assertTrue(spillCipher.isDestroyed());

        spillCipher.destroy(); // should not throw an exception

        assertFailure(() -> spillCipher.decrypt(encrypted), "Spill cipher already destroyed");
        assertFailure(() -> spillCipher.encrypt(randomBytes(1)), "Spill cipher already destroyed");
    }

    private static void assertFailure(ThrowingRunnable runnable, String expectedErrorMessage)
    {
        PrestoException exception = expectThrows(PrestoException.class, runnable);
        assertEquals(exception.getErrorCode(), GENERIC_INTERNAL_ERROR.toErrorCode());
        assertEquals(exception.getMessage(), "Spill cipher already destroyed");
    }

    private static byte[] randomBytes(int size)
    {
        byte[] bytes = new byte[size];
        ThreadLocalRandom.current().nextBytes(bytes);
        return bytes;
    }
```

pettyjamesm replied:
Sounds good, using this.

mbasmanova commented on presto-main/src/test/java/com/facebook/presto/spiller/TestSpillCipher.java:
```diff
@@ -0,0 +1,80 @@
+
+    private static void testRoundTripData(SpillCipher spillCipher, byte[] data, String message)
+    {
+        assertEquals(data, spillCipher.decrypt(spillCipher.encrypt(data.clone())), message);
```
why `data.clone()` and not just `data`?

pettyjamesm replied:
This prevents the implementations from modifying the contents of the `byte[]` that we then use to `assertEquals` on. Slightly defensive, but preserves the correctness of the assertion.

mbasmanova commented on presto-main/src/test/java/com/facebook/presto/spiller/TestSpillCipherPagesSerde.java:
```diff
@@ -0,0 +1,65 @@
+
+        cipher.destroy();
+
+        PrestoException serializeException = expectThrows(PrestoException.class, () -> serde.serialize(nonEmpty));
```
consider using helper method and checking error message:

```
        assertFailure(() -> serde.serialize(nonEmpty), "Spill cipher already destroyed");
        assertFailure(() -> serde.deserialize(serialized), "Spill cipher already destroyed");
    }

    private static void assertFailure(Assert.ThrowingRunnable runnable, String expectedErrorMessage)
    {
        PrestoException exception = expectThrows(PrestoException.class, runnable);
        assertEquals(exception.getErrorCode(), GENERIC_INTERNAL_ERROR.toErrorCode());
        assertEquals(exception.getMessage(), expectedErrorMessage);
    }
```

pettyjamesm replied:
Will do, thanks.

mbasmanova commented on presto-main/src/test/java/com/facebook/presto/spiller/TestSpillCipherPagesSerde.java:
```diff
@@ -0,0 +1,65 @@
+public class TestSpillCipherPagesSerde
+{
+    @Test
+    public void testSpillerEncryptedPagesSerde()
```
I'd rename it to `test` for simplicity.

pettyjamesm replied:
Makes sense, will do.

mbasmanova commented on presto-docs/src/main/sphinx/admin/spill.rst:
```diff
@@ -65,6 +65,27 @@ Presto treats spill paths as independent disks (see `JBOD
+
+When spill compression is enabled (``spill-compression-enabled`` property in
+:ref:`tuning-spilling`), spilled pages will be compressed if they're sufficiently
+compressible. Enabling this feature can reduce the amount of disk IO at the cost
```
Perhaps, specify what type of compression is being used.

pettyjamesm replied:
I'll mention that the compression works the same way as exchange compression, but don't want to explicitly mention the codec name. It's LZ4 for now, but only because that's hard-coded in the `PagesSerdeFactory`. If someone were to add the ability to configure the compression codec there, my documentation would become incorrect.

pettyjamesm commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/PageCodecMarker.java:
```diff
@@ -13,36 +13,42 @@
 
 import static com.google.common.base.Preconditions.checkArgument;
 
+/**
```
I'll remove the static helpers that are `@VisibleForTesting`. I'll also remove the `.of` method and rename `.clear` to `.unset`. Tests will be slimmed, but I'm going to keep the style of looping over `PageCodec.values()` in the tests instead of hard coding `COMPRESSED` and `ENCRYPTED`. The looping over all possible values strategy prevents someone from adding a new marker type that accidentally uses a conflicting mask with other markers.

pettyjamesm commented:
@mbasmanova Addressed your latest round of comments and split spill compression from spill encryption. Here's the result of the benchmark running on my local machine. Obviouslym standard disclaimers about trusting benchmark data apply- especially since real world performance will depend on the I/O throughput of presto hosts, compressibility of the data, etc.

```
# JMH version: 1.20
# VM version: JDK 1.8.0_202, VM 25.202-b08
# VM invoker: /Library/Java/JavaVirtualMachines/jdk1.8.0_202.jdk/Contents/Home/jre/bin/java
# VM options: -Dfile.encoding=UTF-8
# Warmup: 5 iterations, 2000 ms each
# Measurement: 10 iterations, 500 ms each
# Timeout: 10 min per iteration
# Threads: 1 thread, will synchronize iterations
...
Benchmark                         (compressionEnabled)  (encryptionEnabled)  (pagesCount)  (rowsPerPage)   Mode  Cnt    Score    Error  Units
BenchmarkBinaryFileSpiller.read                  false                false            10          10000  thrpt   30  294.620 ± 24.326  ops/s
BenchmarkBinaryFileSpiller.read                  false                 true            10          10000  thrpt   30  154.744 ±  1.778  ops/s
BenchmarkBinaryFileSpiller.read                   true                false            10          10000  thrpt   30  200.231 ±  2.990  ops/s
BenchmarkBinaryFileSpiller.read                   true                 true            10          10000  thrpt   30  162.602 ±  4.224  ops/s
BenchmarkBinaryFileSpiller.write                 false                false            10          10000  thrpt   30  172.535 ± 11.208  ops/s
BenchmarkBinaryFileSpiller.write                 false                 true            10          10000  thrpt   30   87.705 ±  4.519  ops/s
BenchmarkBinaryFileSpiller.write                  true                false            10          10000  thrpt   30   73.161 ± 13.345  ops/s
BenchmarkBinaryFileSpiller.write                  true                 true            10          10000  thrpt   30   72.708 ±  3.120  ops/s
```

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/PagesSerde.java:
```diff
@@ -58,20 +56,18 @@ public SerializedPage serialize(Page page)
-        int maxCompressedLength = maxCompressedLength(serializationBuffer.size());
-        byte[] compressionBuffer = new byte[maxCompressedLength];
-        int actualCompressedLength = compressor.get().compress(serializationBuffer.slice().getBytes(), 0, serializationBuffer.size(), compressionBuffer, 0, maxCompressedLength);
+        byte[] compressionBuffer = new byte[compressor.get().maxCompressedLength(serializationBuffer.size())];
```
this change seems unrelated; it is preferable to put unrelated changes into separate commits

mbasmanova commented on presto-main/src/test/java/com/facebook/presto/execution/buffer/TestPageCodecMarker.java:
```diff
@@ -0,0 +1,56 @@
+import static org.testng.Assert.assertFalse;
+import static org.testng.Assert.assertTrue;
+
+public class TestPageCodecMarker
```
I find these tests hard to read. The complexity of the test code seems to exceed the complexity of the enum being tested. I think testing the only two possible values (as suggested in an earlier comment) and their combinations is appropriate and ensures that production use cases are covered.

pettyjamesm replied:
I'll add a more specific concrete test method, but feel pretty strongly about the exhaustive checking approach as a long-term safeguard against regressions that may be introduced with new `PageCodecMarker` values. For instance, adding a new marker type but accidentally reusing the bit position from another value or extending the masking logic to include the compression codec alongside the `COMPRESSED` marker might easily introduce conflicting marker sets that wouldn't be explicitly combined in concrete tests.

I agree that it makes the test seem more complex than the `PageCodecMarker` implementation itself, but I've caught enough logical errors using a more "property based testing" style approach that I'm convinced they're worth at least a small complexity cost.

As always, if I haven't convinced you then I'm fine removing it to get the PR merged. I think I've addressed all your other comments.

mbasmanova commented on presto-main/src/test/java/com/facebook/presto/spiller/TestFileSingleStreamSpiller.java:
```diff
@@ -78,6 +101,14 @@ public void testSpill()
         assertEquals(listFiles(spillPath.toPath()).size(), 1);
 
+        // Assert the spill codec flags match the expected configuration
+        try (InputStream is = java.nio.file.Files.newInputStream(listFiles(spillPath.toPath()).get(0))) {
```
static import java.nio.file.Files.newInputStream

pettyjamesm replied:
Done

mbasmanova commented on presto-main/src/test/java/com/facebook/presto/spiller/TestFileSingleStreamSpiller.java:
```diff
@@ -78,6 +101,14 @@ public void testSpill()
+        try (InputStream is = java.nio.file.Files.newInputStream(listFiles(spillPath.toPath()).get(0))) {
+            Iterator<SerializedPage> serializedPages = PagesSerdeUtil.readSerializedPages(new InputStreamSliceInput(is));
+            assertTrue(serializedPages.hasNext(), "at least one page should be successfully read back");
+            SerializedPage readBack = serializedPages.next();
```
readBack sounds strange; reconstructedPage might be better; better yet, simplify to

```
            byte marker = serializedPages.next().getPageCodecMarkers();
            assertEquals(PageCodecMarker.COMPRESSED.isSet(marker), compression);
            assertEquals(PageCodecMarker.ENCRYPTED.isSet(marker), encryption);
```

pettyjamesm replied:
Looks better, will use this instead.

mbasmanova commented on presto-main/src/test/java/com/facebook/presto/spiller/TestFileSingleStreamSpiller.java:
```diff
@@ -64,11 +68,30 @@ public void tearDown()
+        testSpillWithSettings(true);
+    }
+
+    private void testSpillWithSettings(boolean compression)
```
for consistency, consider renaming to assertSpill

pettyjamesm replied:
Renamed.

mbasmanova commented on presto-main/src/test/java/com/facebook/presto/spiller/BenchmarkAESSpillCipher.java:
```diff
@@ -0,0 +1,85 @@
+@Warmup(batchSize = 1, iterations = 1, time = 10, timeUnit = TimeUnit.SECONDS)
+@Measurement(batchSize = 1, iterations = 1, time = 3, timeUnit = TimeUnit.SECONDS)
+@BenchmarkMode(Mode.Throughput)
+public class BenchmarkAESSpillCipher
```
- This benchmark is missing main method. How do you run it? 
- @Param("4196") is unnecessary because there is only one value
- Usually, we create a BenchmarkData class with setup method and add BenchmarkData parameter to benchmark methods like below
- It is helpful to add a @Test method to do some basic validation - it will run on each commit and ensure that benchmark doesn't get terribly broken over time.

```
@State(Scope.Thread)
@Fork(2)
@Warmup(iterations = 10, time = 500, timeUnit = TimeUnit.MILLISECONDS)
@Measurement(iterations = 10, time = 500, timeUnit = TimeUnit.MILLISECONDS)
@OutputTimeUnit(TimeUnit.MILLISECONDS)
@BenchmarkMode(Mode.Throughput)
public class BenchmarkAESSpillCipher
{
    @State(Scope.Thread)
    public static class BenchmarkData
    {
        private final AesSpillCipher spillCipher = new AesSpillCipher();
        private byte[] plainText;
        private byte[] cipherText;

        @Setup
        public void setup()
                throws IOException
        {
            this.plainText = createRandomBytes(4196);
            this.cipherText = spillCipher.encrypt(plainText);
        }
    }

    @Benchmark
    public byte[] decryptArray(BenchmarkData data)
    {
        return data.spillCipher.decrypt(data.cipherText);
    }

    @Benchmark
    public byte[] encryptArray(BenchmarkData data)
    {
        return data.spillCipher.encrypt(data.plainText);
    }

    @Benchmark
    public ByteBuffer decryptBuffer(BenchmarkData data)
    {
        return data.spillCipher.decrypt(ByteBuffer.wrap(data.cipherText));
    }

    @Benchmark
    public ByteBuffer encryptBuffer(BenchmarkData data)
    {
        return data.spillCipher.encrypt(ByteBuffer.wrap(data.plainText));
    }

    private static byte[] createRandomBytes(int size)
    {
        byte[] bytes = new byte[size];
        ThreadLocalRandom.current().nextBytes(bytes);
        return bytes;
    }

    @Test
    public void verify()
            throws IOException
    {
        BenchmarkData data = new BenchmarkData();
        data.setup();
        BenchmarkAESSpillCipher benchmark = new BenchmarkAESSpillCipher();
        benchmark.decryptArray(data);
        benchmark.decryptBuffer(data);
        benchmark.encryptArray(data);
        benchmark.encryptBuffer(data);
    }

    public static void main(String[] args)
            throws RunnerException
    {
        Options options = new OptionsBuilder()
                .verbosity(VerboseMode.NORMAL)
                .include(".*" + BenchmarkAESSpillCipher.class.getSimpleName() + ".*")
                .build();
        new Runner(options).run();
    }
}
```

Here are the results from above:

```
# Run complete. Total time: 00:01:31

Benchmark                               Mode  Cnt    Score    Error   Units
BenchmarkAESSpillCipher.decryptArray   thrpt   20  124.728 ±  9.553  ops/ms
BenchmarkAESSpillCipher.decryptBuffer  thrpt   20  135.322 ± 10.958  ops/ms
BenchmarkAESSpillCipher.encryptArray   thrpt   20   85.980 ±  3.884  ops/ms
BenchmarkAESSpillCipher.encryptBuffer  thrpt   20   81.870 ±  5.894  ops/ms
```

The errors are pretty high (~10%). The benchmark setting may need to be tuned. However, it is not clear to me whether this benchmark is needed at all. If not, feel free to just delete.

I also noticed that benchmark results you posted have high error (10%). I don't think this is due to any of your changes, but it says that the benchmark needs some fixing to be useful.

pettyjamesm replied:
I think removing the benchmark is fine. We carried it internally because the original spill encryption PR that is superseded by this one was significantly less performant. I don't think it has value since it's the sole implementation now.

mbasmanova commented:
@pettyjamesm James, thanks for the updates. Everything looks good expect for I have a question about the `key`.

pettyjamesm commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/PageCodecMarker.java:
```diff
@@ -24,7 +24,8 @@
 {
-    COMPRESSED(1);
+    COMPRESSED(1),
+    ENCRYPTED(2);
```
Ok, moved to first commit.

pettyjamesm commented on presto-main/src/test/java/com/facebook/presto/spiller/TestFileSingleStreamSpiller.java:
```diff
@@ -68,17 +68,26 @@ public void tearDown()
 
-    private void testSpillWithSettings(boolean compression)
+    @Test
+    public void testSpillEncryption()
```
Will do.

pettyjamesm commented on presto-main/src/test/java/com/facebook/presto/spiller/TestSpillCipherPagesSerde.java:
```diff
@@ -0,0 +1,70 @@
+        SpillCipher cipher = new AesSpillCipher();
+        PagesSerde serde = new TestingPagesSerdeFactory().createPagesSerdeForSpill(Optional.of(cipher));
+        List<Type> types = ImmutableList.of(VARCHAR);
+        Page empty = new Page(VARCHAR.createBlockBuilder(null, 0).build());
```
Done.

pettyjamesm commented on presto-main/src/test/java/com/facebook/presto/spiller/TestSpillCipherPagesSerde.java:
```diff
@@ -0,0 +1,70 @@
+        BlockBuilder blockBuilder = VARCHAR.createBlockBuilder(null, 2);
+        VARCHAR.writeString(blockBuilder, "hello");
+        VARCHAR.writeString(blockBuilder, "world");
+        Page nonEmpty = new Page(blockBuilder.build());
```
Will do.

pettyjamesm commented:
@mbasmanova I think I've addressed the latest rounds of comments. I added a small change to the original logic [in PagesSerde](https://github.com/prestodb/presto/pull/12617/commits/9bbfb6cb4c6760f4d75091f4d83ee26de27875c5#diff-d069ed44681213694a65b86ebaf27754R82) to copy non-encrypted slices into a new buffer regardless of whether it was compressed or not. This preserves the original logic for non-encrypted, non-compressed pages which may not be appropriately sized for their backing buffer because of `DynamicSliceOutput` behavior.

mbasmanova commented on presto-main/src/test/java/com/facebook/presto/execution/buffer/TestPageCodecMarker.java:
```diff
@@ -0,0 +1,88 @@
+    }
+
+    @Test
+    public void testPageMarkerSanity()
```
- This test tests `isSet` method, hence, let's rename it to `testIsSet`.
- s/m/marker
- drop messages from asserts; these tests are trivial and messages introduce unnecessary visual clutter


```
    @Test
    public void testIsSet()
    {
        assertEquals((byte) 0, PageCodecMarker.none());

        for (PageCodecMarker marker : PageCodecMarker.values()) {
            assertFalse(marker.isSet(PageCodecMarker.none()));
            assertTrue(marker.isSet(marker.set(PageCodecMarker.none())));
            assertFalse(marker.isSet(marker.unset(marker.set(PageCodecMarker.none()))));

            for (PageCodecMarker other : PageCodecMarker.values()) {
                assertEquals(other == marker, other.isSet(marker.set(PageCodecMarker.none())));
            }
        }
    }

    @Test
    public void testSummaryString()
    {
        assertEquals(PageCodecMarker.toSummaryString(PageCodecMarker.none()), "NONE");
        for (PageCodecMarker marker : PageCodecMarker.values()) {
            assertEquals(PageCodecMarker.toSummaryString(marker.set(PageCodecMarker.none())), marker.name());
        }

        byte allMarkers = PageCodecMarker.none();
        Arrays.stream(PageCodecMarker.values()).forEach(marker -> marker.set(allMarkers));
        
        String allMarkersSummary = PageCodecMarker.toSummaryString(allMarkers);
        for (PageCodecMarker marker : PageCodecMarker.values()) {
            assertTrue(allMarkersSummary.contains(marker.name()));
        }
    }
```

pettyjamesm replied:
Will do, unfortunately the usage of `Arrays.stream(PageCodecMarker.values()).forEach(...)` doesn't work since allMarkers needs to be reassigned which is prohibited for lambda closures (closed-over values must be final / effectively final). I'll build it in the loop above but otherwise keep this code.

mbasmanova commented:
@pettyjamesm James, I have some comments re: TestPageCodecMarker. Other than that it looks good.

mbasmanova commented:
I'll merge this once Travis is green.

pettyjamesm commented:
Tests are green except for the `HIVE_TESTS=true` job, which was still trying to load docker containers when I checked on it at around the 40 minute mark. The job eventually loaded the container but timed out before completing.

mbasmanova commented:
@pettyjamesm Thank you for the contribution!
